### DISCUSSION TITLE
Easy recursive assamble for affine decomposable operators

### DISCUSSION THREAD

**Discussion by mfeuerle** [pymor/pymor#1501] [Ideas]
I am currently working on more complicated operator structures (not just a simple `LincompOperator` of a bunch of `NumpyMatrixOperators`), and implementing my own tensor product / kronecker product operator.

My case all involved quantities are all fully affine decomposable and since the affine decomposition is a elementary assumption for RBM i want to make use of it, e.g. using the `SimpleCoerciveRBReductor`.
Unfortunately, to ensure that the `Operator` is affinely decomposed, it is checked (at least in `SimpleCoerciveRBReductor`) if the `Operator` is a `LincompOperator` and all `Operators` belonging to it are not parametric.
This, of course, is a valid approach, but not very convenient, since it is very restrictive. As an example consider the two affine decomposed `Operators`

<img src="https://latex.codecogs.com/svg.latex?A(u;\mu)=\sum_{q=1}^{Q_A}\vartheta^A_q(\mu)A_q(u)"  />
<img src="https://latex.codecogs.com/svg.latex?B(u;\mu)=\sum_{q=1}^{Q_B}\vartheta^B_q(\mu)B_q(u)"  />

then, a `LincompOperator` (or just die addition) of those two is again affine decomposable with

<img src="https://latex.codecogs.com/svg.latex?C(u;\mu):=\vartheta_A(\mu)A(u;\mu)+\vartheta_B(\mu)B(u;\mu)=\sum_{q=1}^{Q_A}\vartheta_A(\mu)\vartheta^A_q(\mu)A_q(u)+\sum_{q=1}^{Q_B}\vartheta_B\vartheta^B_q(\mu)B_q(u)"  />
<img src="https://latex.codecogs.com/svg.latex?=:\sum_{q=1}^{Q_C}\vartheta^C_q(\mu)C_q(u),\qquad&space;Q_C=Q_A+Q_B"  />

Furthermore, if `A` is linear, then the `ConcatenationOperator` of those is also affine decomposable with

<img src="https://latex.codecogs.com/svg.latex?D(u;\mu):=(A(\mu)\circ&space;B(\mu))(u)=A(B(u;\mu);\mu)=\sum_{q=1}^{Q_A}\sum_{q'=1}^{Q_B}\vartheta^A_q(\mu)\vartheta^B_{q'}(\mu)A_q(B_{q'}(u))"  />
<img src="https://latex.codecogs.com/svg.latex?=:\sum_{q=1}^{Q_D}\vartheta^D_q(\mu)D_q(u),\qquad&space;Q_D=Q_AQ_B"  />

But both `Operators` are not registered as suitable for `SimpleCoerciveRBReductor`, because the `Operators` `A` and `B` of `C` are parametric, and the `Operator` `D` is not a `LincompOperator`.

**Therefore, my suggestion:**
Since a affine decomposition is a key element for RBM and probably not only needed for `SimpleCoerciveRBReductor`, i would suggest, to extend the general `Operator` interface with two new functions:

```
Operator.is_affine_decomposable()
# returns True, if the Operator can be represented as affine decomposition
```
```
Operator.assamble_affine_decomposable()
# returns a affine decomposition of the Operator if possible (most likely as LincombOperator?)
```

Those two functions should recursively check, if all involved structures are affine decomposable and assemble the affine representation of the `Operator` if needed.
This would generalise this importand property to a consistent structure instead of requiring and manually checking if a `Operator` is already given as affine decomposition.

**Comment by sdrave** [comment]
Hi @moro1111, `SimpleCoerciveRBReductor` only makes this assumption on the structure of the operator to assemble the error estimator. `CoerciveRBReductor` uses `algorithms.image.estimate_image` to build an orthonormal basis for the image of the residual operator and then performs a Petrov-Galerkin projection to assemble the an mathematically equivalent error estimator. Both `estimate_image` and `project` should be smart enough to handle the constructions you mentioned (otherwise I'd consider it as a bug).

Alternatively: In `algorithms.simplify` you can find an `expand` method, which should expand concatenations of `LincombOperators` and also flatten hierarchies of `LincombOperators`.

Does that work for you?

**Reply by mfeuerle** [reply]
Oh, nice, i did not realize that. The `CollectOperatorRangeRules` class is not very easy to look through in a first glance.

**Reply by ftalbrecht** [reply]
If that works for you, could you mark it as an answer?

**Reply by mfeuerle** [reply]
Since this was written in "Ideas" and not in "Q&A" there is no opportunity to mark something as answer i suppose.
